### DISCUSSION TITLE
Custom Component support for Uno Figma

### DISCUSSION THREAD

**Discussion by tylandercasper** [unoplatform/uno#10136] [Ideas]
Right now the Figma tool only exports components that are predefined in the document. 

I'd like to have custom components so that common UI code can be shared. 

Unfortunately the figma plugin is not currently open source, but I would be willing to donate some programming time to help make this happen. 

Background:
I was super excited about the Uno Figma plugin as the designer/developer handoff is the brunt of our developer efforts right now and a shared UI codebase would solve an enormous about of trouble and be very good for long term support. 

Unfortunately as it is right now, that xaml essentially denests all of the components the user created in Figma and exports flat xaml that uses it's specific controls. It's certainly better than nothing and a handy tool for figuring out how to make something look right, but a far cry from what's needed in to be in a formal workflow. 

Custom Components is a game changer and a huge differentiator from other platforms since the code you're outputting is actually intelligent and reusable. 

Implementation ideas:
For creating custom Components: 
 A base UnoComponent that works like the UserControl in Uno. 
  A user could make the component in Figma with the controls that are particular to it and than drag that component onto the various pages that need it. 
 The ability to right click on the component it and export it's xaml. 

For referencing custom components: 
 I imagine this would be pretty easy. Instead of saying "Is component in predefined doc", change it to "does component inherent from an item in the predefined doc. Than just spit out the CustomName.

**Comment by sasakrsmanovic** [comment]
Hello. Do you mind reaching out to me at sasha@platform.uno so we can explore this idea a bit deeper. Thanks

**Reply by sasakrsmanovic** [reply]
also a cc to @carldebilly
